R/filters.R

```diff
@@ -74,3 +119,174 @@ filter_by_meta <- function(.data, .query) {
 
   return(list(data = filtered_data, meta = filtered_meta))
 }
+
+filter_by_repertoire <- function(.data, .query) {
+  filtered_data <- .data$data
+  for (i in seq_along(names(.query))) {
+    name <- names(.query)[i]
+    key_query <- .query[[name]]
+    query_type <- key_query[[1]]
+    query_args <- key_query[-1]
+    if (length(filtered_data) > 0) {
+      counts <- switch(name,
+        n_clonotypes = lapply(filtered_data, nrow),
+        n_clones = lapply(filtered_data, function(sample) {
+          sum(sample$Clones)
+        }),
+        stop(paste0(
+          "Wrong filter key for \"by.repertoire\": \"", name,
+          "\"! Available keys: \"n_clonotypes\", \"n_clones\"."
+        ))
+      )
+
+      if (query_type == "lessthan") {
+        filtered_data <- filtered_data[
+          counts < as_numeric_or_fail(query_args)
+        ]
+      } else if (query_type == "morethan") {
+        filtered_data <- filtered_data[
+          counts > as_numeric_or_fail(query_args)
+        ]
+      } else if (query_type == "interval") {
+        from <- as_numeric_or_fail(query_args[[1]])
+        to <- as_numeric_or_fail(query_args[[2]])
+        filtered_data <- filtered_data[
+          counts >= from & counts < to
+        ]
+      }
+
+      if (nrow(filtered_data) == 0) {
+        warning(paste0("Filter by key \"", name, "\" removed all remaining samples!"))
+      }
+    }
+  }
+  filtered_meta <-
+    if ("meta" %in% names(.data)) {
+      filter(.data$meta, Sample %in% names(filtered_data))
+    } else {
+      warning("No metadata in input dataset!")
+      tibble()
+    }
+
+  return(list(data = filtered_data, meta = filtered_meta))
+}
+
+filter_by_clonotype <- function(.data, .query, .match) {
+  filtered_data <- .data$data
+  for (i in seq_along(names(.query))) {
+    name <- names(.query)[i]
+    column_query <- .query[[name]]
+    query_type <- column_query[[1]]
+    query_args <- column_query[-1]
+    for (j in seq_along(names(filtered_data))) {
+      sample_name <- names(filtered_data)[j]
+      sample <- filtered_data[[sample_name]]
+      if (!(name %in% names(sample))) {
+        stop(paste0("Column \"", name, "\" not found in sample \"", sample_name, "\"."))
+      }
+      if (nrow(sample) == 0) {
+        warning(paste0("Sample \"", sample_name, "\" was empty, removed!"))
```

Comment:
Why we should remove empty samples?

Reply:
I think we must not have empty samples in input data, so this indicates that something is wrong with the data. Empty samples are not usable, that's why I remove them in case they are completely filtered out (and produce warninig, as we discussed on sync).